File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int>dq;
        int n=nums.size(),addIdx,rmvIdx;
        vector<int>result;
        for(int i=0;i<k;i++){
            while(dq.size()>0 && dq.back()<nums[i]){
                dq.pop_back();
            }
            dq.push_back(nums[i]);
        }
        result.push_back(dq.front());
        for(int i=k; i<n; i++){
            addIdx = i;
            rmvIdx = i-k;
                        
            // expand the window
            while(dq.size()>0 && dq.back()<nums[addIdx]){
                dq.pop_back();
            }
            
            dq.push_back(nums[addIdx]);
            
            // compress the window
            if(dq.front() == nums[rmvIdx]){
                dq.pop_front();
            }
            
            result.push_back(dq.front());
        }
        return result;
    }
};

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> d = new ArrayDeque<>();
        int[] res = new int[nums.length - k + 1];
        int ri = 0;
        for (int i = 0; i < nums.length; i++) {
            if (!d.isEmpty() && d.peek() == i - k) {
                d.pop();
            }
            while (!d.isEmpty() && nums[d.peekLast()] < nums[i]) {
                d.pollLast();
            }
            d.addLast(i);
            if (i >= k - 1) {
                res[ri++] = nums[d.peek()];
            }
        }
        return res;
    }
}